tests/ludwig/utils/test_class_balancing.py

```diff
@@ -0,0 +1,178 @@
+import contextlib
+
+import numpy as np
+import pandas as pd
+import pytest
+import ray
+
+from ludwig.backend import create_ray_backend
+from ludwig.backend.base import LocalBackend
+from ludwig.backend.ray import RayBackend
+from ludwig.constants import NAME, OUTPUT_FLAG, PROC_COLUMN
+from ludwig.data.preprocessing import balance_data
+from tests.integration_tests.utils import spawn
+
+DFS = {
+    "test_df_1": pd.DataFrame(
+        {
+            "Index": np.arange(0, 200, 1),
+            "random_1": np.random.randint(0, 50, 200),
+            "random_2": np.random.choice(["Type A", "Type B", "Type C", "Type D"], 200),
+            "Label": np.concatenate((np.zeros(180), np.ones(20))),
+            "split": np.zeros(200),
+        }
+    )
+}
+
+CONFIGS = {
+    "test_config_1": {
+        "input_features": [
+            {"name": "Index", "column": "Index", "type": "numerical", "output_flag": False},
+            {"name": "random_1", "column": "random_1", "type": "numerical", "output_flag": False},
+            {"name": "random_2", "column": "random_2", "type": "numerical", "output_flag": False},
+        ],
+        "output_features": [{"name": "Label", "column": "Label", "type": "binary", "output_flag": True}],
+        "preprocessing": {},
+    }
+}
+
+
+@contextlib.contextmanager
+def ray_start(num_cpus=2, num_gpus=None):
+    res = ray.init(
+        num_cpus=num_cpus,
+        num_gpus=num_gpus,
+        include_dashboard=False,
+        object_store_memory=150 * 1024 * 1024,
+    )
+    try:
+        yield res
+    finally:
+        ray.shutdown()
+
+
+@spawn
+def run_test_balance_data_ray(
+    input_df,
+    input_features,
+    preprocessing_parameters,
+    target,
+    num_cpus=2,
+    num_gpus=None,
+):
+    with ray_start(num_cpus=num_cpus, num_gpus=num_gpus):
+        backend = create_ray_backend()
+        input_df = backend.df_engine.from_pandas(input_df)
+        test_df = balance_data(input_df, input_features, preprocessing_parameters, backend)
+
+        new_class_balance = round(
+            test_df[target].value_counts().compute()[test_df[target].value_counts().compute().idxmin()]
+            / test_df[target].value_counts().compute()[test_df[target].value_counts().compute().idxmax()],
+            1,
+        )
+
+        assert input_df.compute().shape == (200, 6)
+        assert input_df[input_df["Label"] == 1].compute().shape == (20, 6)
+        assert input_df[input_df["Label"] == 0].compute().shape == (180, 6)
+        assert test_df.compute().shape == (270, 6)
+        assert isinstance(backend, RayBackend)
+        assert new_class_balance == 0.5
```

Comment:
Could we pass in `new_class_balance` as a variable to this test helper functions and add additional test cases for class balance percentages other than `0.5`.

Reply:
Sounds good, I'll test for 0.25 and 0.75 as well